Student: submit using submission link: give clearer indication of success #5979

HirdayGupta opened the pull request:
Fixes #5979 

**Outline of solution**:
If either `studentFeedbackSubmissionEditPage` or the `instructorFeedbackSubmissionEditPage` load with a success status message, a modal is displayed in addition to the regular status message div to clearly indicate a successful submission. 

**Screenshot of modal:**
<img width="677" alt="screen shot 2017-07-09 at 4 36 31 am" src="https://user-images.githubusercontent.com/18088721/27989824-3bb38420-6460-11e7-8207-e52cea4e9869.png">

**Note:** I have noticed that the tick icon pushes the modal title to a new line in the modal header. This will be fixed in #7708.


**Question regarding approach (Senior dev input required):**
- As of now, this is only for unregistered users (who submit using submission link). Would it be necessary to add this in for registered students as well? (This will result in redirecting registered students to the feedback submission edit page as well after a successful submission instead of the current behavior of redirecting them to the student home page).

**Todo:**
- ~~Add test for visibility of modal.~~ Added this test in https://github.com/TEAMMATES/teammates/pull/7710/commits/08cc758242820c9f5daea0afa002bf28f8fb7429.

damithc commented:
> As of now, this is only for unregistered users (who submit using submission link). Would it be necessary to add this in for registered students as well? (This will result in redirecting registered students to the feedback submission edit page as well after a successful submission instead of the current behavior of redirecting them to the student home page).

We can use the same approach for for both type of students.  There is a higher probability of registered students wanting to tweak the submission and resubmit rather than go back to home page.

HirdayGupta commented:
@damithc changed the behavior to stay in edit page even after submitting the form for registered students also. Note that in the `createSpecificRedirectResult()` method responsible for deciding which page to redirect to, there was a provision to return to the edit page if there was an error. Since we are now returning to the same page whether or not there is an error, I have removed the if condition and modified the comment accordingly in d36a74a. Hoping this was the right call.

Ready for review. (Will update tests after review of feature).

damithc commented:
> Hoping this was the right call.

Yes, that should be fine behavior-wise. I hope the model style is different when there is an error?

HirdayGupta commented:
@damithc As far as I can see, no modal is displayed when there is an error. Just an `alert-danger` status message with the appropriate error message(s) (as set by `statusMessagesToUser`). The modal I have made in this PR will ONLY display if the responses had no errors and a success message was returned from the server. This is implemented by the following code:

```JavaScript
const SUCCESS_STATUS_MESSAGE = '.alert-success.statusMessage';
function hasSuccessMessage() {
    return $(SUCCESS_STATUS_MESSAGE).length;
}
```

The above code returns `1` only if there is a success message, and so the modal is only displayed if there is a success message.

damithc commented:
> @damithc As far as I can see, no modal is displayed when there is an error
It's fine I guess.

HirdayGupta commented:
@damithc I am now working on adding a test for modal and fixing the existing build. Let me know if any changes needed in code or current behaviour.

damithc commented:
> @damithc I am now working on adding a test for modal and fixing the existing build.

Go ahead @HirdayGupta

HirdayGupta commented:
**Summary of changes**
- Fixed build in commits d5e6658 and 1063a89 (In this, also added tests for modal visibility in `InstructorEditStudentFeedbackPageUiTest`)

- Added test for modal visiblity in `StudentFeedbackSubmitPageUiTest`

**Ready for review**

HirdayGupta commented:
Dropping a note that the misaligned `success icon` problem will go away after #7774 is merged.

HirdayGupta commented:
@whipermr5 Perhaps this can be changed to `s.ToReview`? #7774 Seems to be almost done and senoir devs can always wait before merging, but this PR has no direct relation to #7774.

leeyimin commented on src/main/java/teammates/ui/controller/StudentFeedbackSubmissionEditSaveAction.java:
```diff
@@ -75,20 +75,16 @@ protected RedirectResult createSpecificRedirectResult() {
-            result.responseParams.put(Const.ParamsNames.COURSE_ID, student.course);
-            result.responseParams.put(Const.ParamsNames.FEEDBACK_SESSION_NAME,
-                                      getRequestParamValue(Const.ParamsNames.FEEDBACK_SESSION_NAME));
+        // Return to student home page if user is registered
```
Update the comment.

leeyimin commented on src/test/java/teammates/test/cases/browsertests/InstructorEditStudentFeedbackPageUiTest.java:
```diff
@@ -115,6 +117,7 @@ private void testDeleteResponse() {
                                           "student1InIESFPTCourse@gmail.tmt",
                                           "student1InIESFPTCourse@gmail.tmt");
         assertNull(fr);
+        submitPage.waitForAndDismissAlertModal();
```
Move this line to before line 110?

leeyimin commented on src/test/java/teammates/test/cases/browsertests/InstructorEditStudentFeedbackPageUiTest.java:
```diff
@@ -75,6 +75,7 @@ private void testEditResponse() throws Exception {
                                           "student1InIESFPTCourse@gmail.tmt");
 
         assertEquals("<p>Good design</p>", fr.getResponseDetails().getAnswerString());
+        submitPage.waitForAndDismissAlertModal();
```
Move this line to before line 71?

HirdayGupta commented:
@leeyimin All done. Ready for review! :)

leeyimin commented on src/test/java/teammates/test/cases/browsertests/InstructorEditStudentFeedbackPageUiTest.java:
```diff
@@ -67,6 +67,7 @@ private void testEditResponse() throws Exception {
 
         submitPage.clickSubmitButton();
         submitPage.verifyStatus(Const.StatusMessages.FEEDBACK_RESPONSES_SAVED);
+        submitPage.waitForAndDismissAlertModal();
```
Could create a method in FeedbackSubmitPage to verify that the correct alert modal (message, style class) is created.

HirdayGupta commented:
@leeyimin Added in the test for the correct modal. Ready for review! :)

HirdayGupta commented:
Leaving a note that now that #7774 is merged, the modal is looking normal 🎉

Screenshot:

<img width="643" alt="screen shot 2017-07-26 at 3 47 24 pm" src="https://user-images.githubusercontent.com/18088721/28616283-d5cd4ad4-7219-11e7-85d0-de86c240b63f.png">

HirdayGupta commented:
@leeyimin My apologies, I didn't notice the failing test (Ran the wrong test locally before committing 😅) That being said, I am not entirely sure why the test is failing here. The exception is an `ElementNotVisibleException` but the element that it is trying to fill, the text box, is very much visible when I tried to debug using the `Thread.sleep()`. It seems strange that the same test method should work for the StudentFeedbackSubmitPageUiTest but not the InstructorEditStudentFeedbackPageUiTest.

leeyimin commented:
@HirdayGupta add `waitForModalToDisappear` in your verify method.

HirdayGupta commented:
Thanks @leeyimin! 
@thenaesh Awaiting review :)

whipermr5 commented:
@thenaesh Reminder

HirdayGupta commented:
@thenaesh Reminder.

HirdayGupta commented:
@thenaesh Please review
